File: go/indicators/tusharchande/aroon/aroon.go
package aroon

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

// Aroon is Tushar Chande's Aroon indicator.
//
// The Aroon indicator measures the number of periods since the highest high
// and lowest low within a lookback window. It produces three outputs:
//   - Up: 100 * (Length - periods since highest high) / Length
//   - Down: 100 * (Length - periods since lowest low) / Length
//   - Osc: Up - Down
//
// The indicator requires bar data (high, low). For scalar, quote, and
// trade updates, the single value substitutes for both.
//
// Reference:
//
// Chande, Tushar S. (1995). "The New Technical Trader". John Wiley & Sons.
type Aroon struct {
	mu sync.RWMutex

	length int
	factor float64

	// Circular buffers for high and low values (size = length+1).
	highBuf     []float64
	lowBuf      []float64
	bufferIndex int
	count       int

	// Tracked indices of highest high and lowest low (absolute indices).
	highestIndex int
	lowestIndex  int

	up     float64
	down   float64
	osc    float64
	primed bool

	mnemonic string
}

// NewAroon returns an instance of the indicator created using supplied parameters.
func NewAroon(p *AroonParams) (*Aroon, error) {
	const (
		invalid   = "invalid aroon parameters"
		fmts      = "%s: %s"
		minLength = 2
	)

	if p.Length < minLength {
		return nil, fmt.Errorf(fmts, invalid, "length should be greater than 1")
	}

	mnemonic := fmt.Sprintf("aroon(%d)", p.Length)
	windowSize := p.Length + 1

	return &Aroon{
		length:   p.Length,
		factor:   100.0 / float64(p.Length), //nolint:mnd
		highBuf:  make([]float64, windowSize),
		lowBuf:   make([]float64, windowSize),
		up:       math.NaN(),
		down:     math.NaN(),
		osc:      math.NaN(),
		mnemonic: mnemonic,
	}, nil
}

// IsPrimed indicates whether the indicator is primed.
func (a *Aroon) IsPrimed() bool {
	a.mu.RLock()
	defer a.mu.RUnlock()

	return a.primed
}

// Metadata describes the output data of the indicator.
func (a *Aroon) Metadata() core.Metadata {
	desc := "Aroon " + a.mnemonic

	return core.BuildMetadata(
		core.Aroon,
		a.mnemonic,
		desc,
		[]core.OutputText{
			{Mnemonic: a.mnemonic + " up", Description: desc + " Up"},
			{Mnemonic: a.mnemonic + " down", Description: desc + " Down"},
			{Mnemonic: a.mnemonic + " osc", Description: desc + " Oscillator"},
		},
	)
}

// Update updates the indicator given the next bar's high and low values.
// Returns Up, Down, and Osc.
func (a *Aroon) Update(high, low float64) (float64, float64, float64) {
	if math.IsNaN(high) || math.IsNaN(low) {
		return math.NaN(), math.NaN(), math.NaN()
	}

	a.mu.Lock()
	defer a.mu.Unlock()

	windowSize := a.length + 1
	today := a.count

	// Store in circular buffer.
	pos := a.bufferIndex
	a.highBuf[pos] = high
	a.lowBuf[pos] = low
	a.bufferIndex = (a.bufferIndex + 1) % windowSize
	a.count++

	// Need at least length+1 bars (indices 0..length).
	if a.count < windowSize {
		return a.up, a.down, a.osc
	}

	trailingIndex := today - a.length

	if a.count == windowSize {
		// First time: scan entire window to find highest/lowest.
		a.highestIndex = trailingIndex
		a.lowestIndex = trailingIndex

		for i := trailingIndex + 1; i <= today; i++ {
			bufPos := i % windowSize

			if a.highBuf[bufPos] >= a.highBuf[a.highestIndex%windowSize] {
				a.highestIndex = i
			}

			if a.lowBuf[bufPos] <= a.lowBuf[a.lowestIndex%windowSize] {
				a.lowestIndex = i
			}
		}
	} else {
		// Subsequent: optimized update.
		// Check if tracked indices fell out of window.
		if a.highestIndex < trailingIndex {
			a.highestIndex = trailingIndex

			for i := trailingIndex + 1; i <= today; i++ {
				bufPos := i % windowSize
				if a.highBuf[bufPos] >= a.highBuf[a.highestIndex%windowSize] {
					a.highestIndex = i
				}
			}
		} else if high >= a.highBuf[a.highestIndex%windowSize] {
			a.highestIndex = today
		}

		if a.lowestIndex < trailingIndex {
			a.lowestIndex = trailingIndex

			for i := trailingIndex + 1; i <= today; i++ {
				bufPos := i % windowSize
				if a.lowBuf[bufPos] <= a.lowBuf[a.lowestIndex%windowSize] {
					a.lowestIndex = i
				}
			}
		} else if low <= a.lowBuf[a.lowestIndex%windowSize] {
			a.lowestIndex = today
		}
	}

	a.up = a.factor * float64(a.length-(today-a.highestIndex))
	a.down = a.factor * float64(a.length-(today-a.lowestIndex))
	a.osc = a.up - a.down

	if !a.primed {
		a.primed = true
	}

	return a.up, a.down, a.osc
}

// UpdateScalar updates the indicator given the next scalar sample.
func (a *Aroon) UpdateScalar(sample *entities.Scalar) core.Output {
	v := sample.Value
	up, down, osc := a.Update(v, v)

	output := make([]any, 3) //nolint:mnd
	output[0] = entities.Scalar{Time: sample.Time, Value: up}
	output[1] = entities.Scalar{Time: sample.Time, Value: down}
	output[2] = entities.Scalar{Time: sample.Time, Value: osc}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (a *Aroon) UpdateBar(sample *entities.Bar) core.Output {
	up, down, osc := a.Update(sample.High, sample.Low)

	output := make([]any, 3) //nolint:mnd
	output[0] = entities.Scalar{Time: sample.Time, Value: up}
	output[1] = entities.Scalar{Time: sample.Time, Value: down}
	output[2] = entities.Scalar{Time: sample.Time, Value: osc}

	return output
}

// UpdateQuote updates the indicator given the next quote sample.
func (a *Aroon) UpdateQuote(sample *entities.Quote) core.Output {
	v := (sample.Bid + sample.Ask) / 2 //nolint:mnd

	return a.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateTrade updates the indicator given the next trade sample.
func (a *Aroon) UpdateTrade(sample *entities.Trade) core.Output {
	return a.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: sample.Price})
}

File: ts/indicators/tushar-chande/aroon/aroon.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { AroonParams } from './params';

/** Function to calculate mnemonic of an __Aroon__ indicator. */
export const aroonMnemonic = (params: AroonParams): string => `aroon(${params.length})`;

/**
 * Tushar Chande's Aroon indicator.
 *
 * The Aroon indicator measures the number of periods since the highest high
 * and lowest low within a lookback window. It produces three outputs:
 *   - AroonUp: 100 * (Length - periods since highest high) / Length
 *   - AroonDown: 100 * (Length - periods since lowest low) / Length
 *   - AroonOsc: AroonUp - AroonDown
 *
 * The indicator requires bar data (high, low). For scalar, quote, and
 * trade updates, the single value substitutes for both.
 *
 * Reference:
 *
 * Chande, Tushar S. (1995). "The New Technical Trader". John Wiley & Sons.
 */
export class Aroon implements Indicator {

  private readonly length_: number;
  private readonly factor: number;

  private readonly highBuf: Float64Array;
  private readonly lowBuf: Float64Array;
  private bufferIndex = 0;
  private count = 0;

  private highestIndex = 0;
  private lowestIndex = 0;

  private up_ = NaN;
  private down_ = NaN;
  private osc_ = NaN;
  private primed_ = false;

  private readonly mnemonic_: string;
  private readonly description_: string;

  constructor(params: AroonParams) {
    const length = Math.floor(params.length);

    if (length < 2) {
      throw new Error('length should be greater than 1');
    }

    this.length_ = length;
    this.factor = 100.0 / length;

    const windowSize = length + 1;
    this.highBuf = new Float64Array(windowSize);
    this.lowBuf = new Float64Array(windowSize);

    this.mnemonic_ = aroonMnemonic(params);
    this.description_ = 'Aroon ' + this.mnemonic_;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean {
    return this.primed_;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.Aroon,
      this.mnemonic_,
      this.description_,
      [
        { mnemonic: this.mnemonic_ + ' up', description: this.description_ + ' Up' },
        { mnemonic: this.mnemonic_ + ' down', description: this.description_ + ' Down' },
        { mnemonic: this.mnemonic_ + ' osc', description: this.description_ + ' Oscillator' },
      ],
    );
  }

  /** Updates the indicator given the next bar's high and low values. Returns [AroonUp, AroonDown, AroonOsc]. */
  public update(high: number, low: number): [number, number, number] {
    if (isNaN(high) || isNaN(low)) {
      return [NaN, NaN, NaN];
    }

    const windowSize = this.length_ + 1;
    const today = this.count;

    // Store in circular buffer.
    const pos = this.bufferIndex;
    this.highBuf[pos] = high;
    this.lowBuf[pos] = low;
    this.bufferIndex = (this.bufferIndex + 1) % windowSize;
    this.count++;

    // Need at least length+1 bars.
    if (this.count < windowSize) {
      return [this.up_, this.down_, this.osc_];
    }

    const trailingIndex = today - this.length_;

    if (this.count === windowSize) {
      // First time: scan entire window.
      this.highestIndex = trailingIndex;
      this.lowestIndex = trailingIndex;

      for (let i = trailingIndex + 1; i <= today; i++) {
        const bufPos = i % windowSize;

        if (this.highBuf[bufPos] >= this.highBuf[this.highestIndex % windowSize]) {
          this.highestIndex = i;
        }

        if (this.lowBuf[bufPos] <= this.lowBuf[this.lowestIndex % windowSize]) {
          this.lowestIndex = i;
        }
      }
    } else {
      // Subsequent: optimized update.
      if (this.highestIndex < trailingIndex) {
        this.highestIndex = trailingIndex;

        for (let i = trailingIndex + 1; i <= today; i++) {
          const bufPos = i % windowSize;
          if (this.highBuf[bufPos] >= this.highBuf[this.highestIndex % windowSize]) {
            this.highestIndex = i;
          }
        }
      } else if (high >= this.highBuf[this.highestIndex % windowSize]) {
        this.highestIndex = today;
      }

      if (this.lowestIndex < trailingIndex) {
        this.lowestIndex = trailingIndex;

        for (let i = trailingIndex + 1; i <= today; i++) {
          const bufPos = i % windowSize;
          if (this.lowBuf[bufPos] <= this.lowBuf[this.lowestIndex % windowSize]) {
            this.lowestIndex = i;
          }
        }
      } else if (low <= this.lowBuf[this.lowestIndex % windowSize]) {
        this.lowestIndex = today;
      }
    }

    this.up_ = this.factor * (this.length_ - (today - this.highestIndex));
    this.down_ = this.factor * (this.length_ - (today - this.lowestIndex));
    this.osc_ = this.up_ - this.down_;

    if (!this.primed_) {
      this.primed_ = true;
    }

    return [this.up_, this.down_, this.osc_];
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const v = sample.value;
    const [up, down, osc] = this.update(v, v);
    const s1 = new Scalar();
    s1.time = sample.time;
    s1.value = up;
    const s2 = new Scalar();
    s2.time = sample.time;
    s2.value = down;
    const s3 = new Scalar();
    s3.time = sample.time;
    s3.value = osc;
    return [s1, s2, s3];
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const [up, down, osc] = this.update(sample.high, sample.low);
    const s1 = new Scalar();
    s1.time = sample.time;
    s1.value = up;
    const s2 = new Scalar();
    s2.time = sample.time;
    s2.value = down;
    const s3 = new Scalar();
    s3.time = sample.time;
    s3.value = osc;
    return [s1, s2, s3];
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = (sample.bid + sample.ask) / 2;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return this.updateScalar(scalar);
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = sample.price;
    return this.updateScalar(scalar);
  }
}
